server/app/views/admin/programs/FormattedPredicateValue.java

```diff
@@ -0,0 +1,142 @@
+  }
+
+  private static FormattedPredicateValue createSingleValue(String value) {
+    return create(Optional.of(value), Optional.empty());
+  }
+
+  private static FormattedPredicateValue createWithSecondValue(
+      String mainValue, String secondValue) {
+    return create(Optional.of(mainValue), Optional.of(secondValue));
+  }
+
+  static FormattedPredicateValue create(Optional<String> mainValue, Optional<String> secondValue) {
+    return new AutoValue_FormattedPredicateValue(mainValue, secondValue);
+  }
+
+  static FormattedPredicateValue fromLeafNode(Optional<LeafOperationExpressionNode> leafNode) {
+    return leafNode.map(FormattedPredicateValue::fromLeafNode).orElse(createEmpty());
+  }
+
+  private static FormattedPredicateValue fromLeafNode(LeafOperationExpressionNode leafNode) {
+    ScalarType scalarType = leafNode.scalar().toScalarType();
+    String value = leafNode.comparedValue().value();
+    OperatorRightHandType predicateType = leafNode.comparedValue().type();
+    switch (scalarType) {
+      case CURRENCY_CENTS:
+        if (predicateType == OperatorRightHandType.PAIR_OF_LONGS) {
+          return parsePairValue(value).map(FormattedPredicateValue::formatCurrency);
+        }
+        return createSingleValue(formatCurrency(value));
+      case DATE:
+        if (predicateType == OperatorRightHandType.LIST_OF_LONGS) {
+          // Backwards compatibility for the original implementation of the AGE_BETWEEN operator
+          // (https://github.com/civiform/civiform/pull/4428) which stored the values as a
+          // LIST_OF_LONGS with exactly two values instead of PAIR_OF_LONGS
+          if (leafNode.operator() == Operator.AGE_BETWEEN) {
+            return parsePairValue(value);
+          }
+          return createSingleValue(formatListOfLongs(value));
+        }
+        if (predicateType == OperatorRightHandType.PAIR_OF_LONGS) {
+          return parsePairValue(value);
+        }
+        if (predicateType == OperatorRightHandType.LONG
+            || predicateType == OperatorRightHandType.DOUBLE) {
+          return createSingleValue(value);
+        }
+        if (predicateType == OperatorRightHandType.PAIR_OF_DATES) {
+          return parsePairValue(value).map(FormattedPredicateValue::formatDate);
+        }
+
+        return createSingleValue(formatDate(value));
+      case LONG:
+        if (predicateType == OperatorRightHandType.LIST_OF_LONGS) {
+          return createSingleValue(formatListOfLongs(value));
+        }
+        if (predicateType == OperatorRightHandType.PAIR_OF_LONGS) {
+          return parsePairValue(value);
+        }
+
+        return createSingleValue(value);
+      case LIST_OF_STRINGS:
+      case STRING:
+        if (predicateType == OperatorRightHandType.LIST_OF_STRINGS) {
+          // Lists of strings are serialized as JSON arrays e.g. "[\"one\", \"two\"]"
+          return createSingleValue(
+              Splitter.on(", ")
+                  // Remove opening and closing brackets
```

Comment:
Thank you but this code is moved from ProgramPredicateConfigureView so I did not write them. :) And actually I would prefer to leave them out because the code seems simple enough to read to me 😅

I looked at creating a utility method but there are already so many tiny utility methods here that it felt like too many layers of indirection.